dispatcher/dispatcher/main.py

```diff
@@ -0,0 +1,67 @@
+import pika
+import json
+import smtplib
+from email.message import EmailMessage
+from jinja2 import Environment, FileSystemLoader, exceptions
+
+"""
+continuously polls(*) the email queue and renders+sends the template on every acked message
+(*) = waits for the queue to push a message onto the app
+"""
+
+environment = Environment(loader=FileSystemLoader("../templates/"))
+
+EMAIL_HOST='mailhog' #FIXME differentiate between dev and prod
+smtpObj = smtplib.SMTP( EMAIL_HOST, 1025 )
+
+RABBIT_HOST='rabbit'
+connection = pika.BlockingConnection(pika.ConnectionParameters(RABBIT_HOST))
+channel = connection.channel()
+
+channel.exchange_declare(exchange='eml',
+                         exchange_type='direct',
+                         durable=True)
+channel.queue_declare(queue='email', durable=True)
+channel.queue_bind(exchange='eml',
+                   queue='email',
+                   routing_key='mail')
+#channel.basic_qos(prefetch_count=1) #TODO: notice that an error processing a message will BLOCK the others from being processed
+
+def send_email(ch, method, properties, body):
+    msg = json.loads(body)
+    try:
+        template = environment.get_template(f"{msg['template']}.jinja2")
+    except exceptions.TemplateNotFound:
+        # TODO: send a notification to someone about adding a template
+        print(f"template {msg['template']}.jinja2 not found")
+        return
+    # TODO: check if there is an auto-delete after 30 minutes for stuck un-acked messages
```

Comment:
I think we need to alert someone to add the template. In rabbit we can see the messages that are un-acked (so it was tried to be processed, but then left there)

Reply:
i still have to add the mechanism that the messages are rerouted to retry at some point in the future (presumably when "a person" adds the missing template

Reply:
If it's just the missing template can we not just reject the message and send an error back that the template does not exist?

Reply:
Same for messages that are missing some values

Reply:
ahhh i thought i replied

the point is that for the dispatcher to reject a message it needs to be synchronous (like it is now) whereas the usage of a message queue is exactly to avoid this synchronicity. Which means that as it currently is, the errors are 2: "missing template" (solvable by human intervention, so the message gets requeued with a delay and operator gets informed to add a new template); and "missing parameter" (unsolvable as there's the need to recreate information, so the message is discarded and a new ticket is created for programmers to fix it). Keep in mind, missing parameter means that there was not a thorough integration testing

Reply:
I agree with Fabri on the lack of integration testing: I would accept that some emails are never sent and that we quickly add the proper template, thanks to an automated ticket/a message somewhere else. (Not sure what the proper error reporting channel can be)

I don't think that we need a mechanism to resend emails that never got sent in a first version, or ideally ever.

How can we easily develop this integration test? Setup one account and 3 events, and send an application, modify, cancel, confirm, and see that we do get one email for each action?

Reply:
hey flooo! 
yes I think it's how you say, bootstrap the 3 containers and send simple API commands after a small provisioning

Reply:
also the proper reporting channel would be slack at the beginning i'd say (we have the small wrapper notify.sh iirc)